src/inputfetcher.h

```diff
@@ -0,0 +1,237 @@
+// Copyright (c) 2024-present The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef BITCOIN_INPUTFETCHER_H
+#define BITCOIN_INPUTFETCHER_H
+
+#include <coins.h>
+#include <sync.h>
+#include <tinyformat.h>
+#include <txdb.h>
+#include <util/threadnames.h>
+
+#include <algorithm>
+#include <iterator>
+#include <set>
+#include <vector>
+
+/**
+ * Input fetcher for fetching inputs from the CoinsDB and inserting
+ * into the CoinsTip.
+ *
+ * The main thread pushes batches of outpoints
+ * onto the queue, where they are fetched by N worker threads. The resulting
+ * coins are pushed onto another queue after they are read from disk. When
+ * the main is done adding outpoints, it starts writing the results of the read
+ * queue to the cache.
+ */
+class InputFetcher
+{
+private:
+    //! Mutex to protect the inner state
+    Mutex m_mutex{};
+
+    //! Worker threads block on this when out of work
+    std::condition_variable m_worker_cv{};
+
+    //! Main thread blocks on this when out of work
+    std::condition_variable m_main_cv{};
+
+    //! The queue of outpoints to be fetched from disk.
+    //! As the order of outpoints doesn't matter, it is used as a LIFO (stack)
+    std::vector<COutPoint> m_outpoints GUARDED_BY(m_mutex){};
+
+    //! The queue of pairs to be written to the cache.
+    std::vector<std::pair<COutPoint, Coin>> m_pairs GUARDED_BY(m_mutex){};
+
+    /**
+     * Number of outpoint fetches that haven't completed yet.
+     * This includes outpoints that are no longer queued, but still in the
+     * worker's own batches.
+     */
+    size_t m_in_flight_fetches_count GUARDED_BY(m_mutex){0};
+
+    //! The maximum number of outpoints to be processed in one batch
+    const size_t m_batch_size;
+
+    //! DB to fetch from.
+    const CCoinsViewDB* m_db{nullptr};
+
+    std::vector<std::thread> m_worker_threads;
+    bool m_request_stop GUARDED_BY(m_mutex){false};
+
+    /** Internal function that does the fetching from disk. */
+    void Loop() noexcept EXCLUSIVE_LOCKS_REQUIRED(!m_mutex)
```

Comment:
We're basically mimicking RocksDB's `MultiGet` here - but prewarming the cache instead in separate get requests, since we can't really take advantage of LevelDB's internals.

Since splitting into buckets isn't trivial and since `MultiGet` seems to rely on C++20 [coroutines](https://en.cppreference.com/w/cpp/language/coroutines) (which wasn't available in 2012 when `CCheckQueue` was written), I'm wondering how much simpler this fetching would be if we had lightweight suspendible threads instead: https://rocksdb.org/blog/2022/10/07/asynchronous-io-in-rocksdb.html#multiget

Reply:
I think it would be similar in complexity, we would still need all the locking mechanisms to prevent multithreaded access.

What would really be great is if we had a similar construction to Rust's `std::sync::mpsc`.

Reply:
> coroutines are just programming paradigm, not magic

That's also what I was counting on! :D

In RocksDB they have [high and low priority work](https://github.com/facebook/rocksdb/wiki/RocksDB-Tuning-Guide#parallelism-options) (I assume that's just added to the front or the back of a background work deque) – this would align well with @furszy's suggestion for mixing different kinds of background work units.

I haven't used the C++ variant of coroutines either, but my thinking was that since they can theoretically yield execution when waiting for IO (and resume later), this would allow threads to focus on other tasks in the meantime. Combined with an appropriate scheduling mechanism (such as a thread pool), we could maximize both CPU and IO usage, if I'm not mistaken.
Instead of each thread handling just one task, it could suspend a coroutine while waiting on IO (e.g., a database fetch) and resume it later, effectively maximizing CPU and IO work without needing to know the exact details of the work.

> If the vectors are thread local

The vector would still be global, but each thread would only access a single bucket (i.e. global vector of vectors, with each thread from the pool writing only to `vector[thread_id]`, which contains a vector of fetched coins).
When all the work is finished, we'd iterate over the global vector and merge the results into the cache on a single thread.
As mentioned, sorting the outpoints before fetching could help improve data locality and reduce lock contention, and the coroutines above would help with work stealing, ensuring that all threads finish roughly at the same time.

Is there anything prohibiting us from doing something like this to minimize synchronization and lock contention during the fetch phase? I understand some synchronization would still be needed during the merge, but this could help reduce global locks and unnecessary synchronization throughout the process.